fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -470,6 +466,23 @@ private ScoredPlan planFilterWithInJoin(@Nonnull PlanContext planContext, @Nonnu
         return null;
     }
 
+    // Get the key expression for the index entries, which includes primary key fields for normal indexes.
+    private KeyExpression indexKeyExpressionForPlan(@Nullable KeyExpression commonPrimaryKey, @Nonnull Index index) {
+        KeyExpression indexKeyExpression = index.getRootExpression();
+        if (indexKeyExpression instanceof KeyWithValueExpression) {
+            indexKeyExpression = ((KeyWithValueExpression) indexKeyExpression).getKeyExpression();
+        }
+        if (commonPrimaryKey != null && indexTypes.getValueTypes().contains(index.getType())) {
```

Comment:
We could put in an additional check of a flag in `RecordQueryPlannerConfiguration` here. That would allow turning this behavior off (giving only the declared fields), if we are concerned about unforeseen consequences.